### PR TITLE
[RFC] Engineering guiding principles

### PR DESCRIPTION
**Pull request by orta**
This tries to make a spiritual successor to the [Artsy objc.io](https://www.objc.io/issues/22-scale/artsy/) as a part of making more of artsy dev culture explicit instead of implicit.

I took a look at some of our larger tech choices and tried to re-frame some of the terms I've used when speaking publicly to be a consistent set of ideals. I need to do a bit more work on declaring that this is something we strive for, rather than perfect being this.

It's worth noting that this is different from the current org-wide culture survey that @williardx is running.

### CODE REVIEW

File: meta/engineering_principles.md
Comment (ashfurrow) [inline meta/engineering_principles.md]: > make decisions on what and how to do engineering**,** we need guiding principals

Add a comma for clarity?

File: meta/engineering_principles.md
Comment (ashfurrow) [inline meta/engineering_principles.md]: These are two sentences spliced together with a comma – I think a dash or semicolon would be better.

File: meta/engineering_principles.md
Comment (ashfurrow) [inline meta/engineering_principles.md]: "business needs that need" sound strange, maybe "business needs that require" or similar?

File: meta/engineering_principles.md
Comment (ashfurrow) [inline meta/engineering_principles.md]: This should be expanded, as it gets explained in more detail in some of the further reading. Repeating and rephrasing the first sentence could provide clarity and help connect it to the second.

File: meta/engineering_principles.md
Comment (ashfurrow) [inline meta/engineering_principles.md]: Another comma splice, maybe replace with a colon?

> different than OSS libs: an app

Comment (ashfurrow) [reply meta/engineering_principles.md]: I'd also considering italicizing `apps` and `libs`.

File: meta/engineering_principles.md
Comment (ashfurrow) [inline meta/engineering_principles.md]: > Instead**,**

Needs a comma.

File: meta/engineering_principles.md
Comment (ashfurrow) [inline meta/engineering_principles.md]: Capitalizing of "Your" should be consistent with other titles.

File: meta/engineering_principles.md
Comment (ashfurrow) [inline meta/engineering_principles.md]: We could link to YouTube videos of JavaScriptures or others (eg: https://www.youtube.com/watch?v=yoGielteT64 )

File: meta/engineering_principles.md
Comment (ashfurrow) [inline meta/engineering_principles.md]: Link to: https://peerlab.community

File: meta/engineering_principles.md
Comment (ashfurrow) [inline meta/engineering_principles.md]: This should match tense with other titles, which are imperative. Reword to "De-silo Engineers"?

File: meta/engineering_principles.md
Comment (ashfurrow) [inline meta/engineering_principles.md]: Same thing: s/Provides/Provide

File: meta/engineering_principles.md
Comment (ashfurrow) [inline meta/engineering_principles.md]: "ideas" -> "idioms" or something a little more substantial? React, for example, is much more than an idea.

**Review by ashfurrow — APPROVED** [review]
Looks great – I would co-sign it all 👍 Just a bunch of nitpicking grammar stuff.

File: meta/engineering_principles.md
Comment (joeyAghion) [inline meta/engineering_principles.md]: I'm not sure this is a viable business practice. Launching projects as open-source means associating a license with the content and potentially accepting external contributions. Making projects private later is not as simple as checking a box and removing the license file; it's unclear whether we even _could_ make outside contributions private without explicitly re-licensing that content. Plus, any external forks of the previously-public project become orphaned or worse--there becomes a new public ["parent" repo](https://help.github.com/articles/what-happens-to-forks-when-a-repository-is-deleted-or-changes-visibility/#changing-a-public-repository-to-a-private-repository) that we don't control.

File: meta/engineering_principles.md
Comment (joeyAghion) [inline meta/engineering_principles.md]: I like the gist of "we're lean, so try to employ the best components from the wider field," but this feels like it's in conflict with also being core contributors to those "best" dependencies. If they're the best, we should contribute but it's unlikely we'll need to "own" them, right?

File: meta/engineering_principles.md
Comment (joeyAghion) [inline meta/engineering_principles.md]: 👍

File: meta/engineering_principles.md
Comment (joeyAghion) [inline meta/engineering_principles.md]: 👍

I'd say "Find ways to _get the value_ of process without additional work."

File: meta/engineering_principles.md
Comment (joeyAghion) [inline meta/engineering_principles.md]: I really relate to this kind of "everyone taking responsibility for the whole group and product" message. We're a lean team so we must maximize the leverage of our individual efforts by sharing openly, etc. I actually think "leverage" is a nice meaningful term that could be incorporated into the title.

**Review by joeyAghion — COMMENTED** [review]
Props for writing these down. Reading and reacting, I wonder if inline comments can lead to "consensus" about such a deep topic, but at least you've started the process.

I selfishly would love to incorporate some of the [Platform team's principles](https://github.com/artsy/potential/blob/master/platform/Principles.md) into this. The most valuable for our work have been:
* "Build for 10x." This helps us decide how much architecture/engineering/planning is too little and how much is too much.
* "Done means done." This makes explicit that we take responsibility for our work product being the _right_ change, performing in production as expected, and being understood and adopted by the necessary stakeholders. I.e., our work is not over when the PR is merged.

**Comment by orta** [comment]
WRT consensus, I figured it might make sense to not class this as something mergeable once some commenting has happened, but do feedback etc for a week and iterate - then do an optional for all meeting about wrapping it up and shipping it

File: meta/engineering_principles.md
Comment (orta) [reply meta/engineering_principles.md]: > it's unclear whether we even could make outside contributions private without explicitly re-licensing that content

Yeah, in those cases we'd have a few interesting choices:

- Debate moving apps into being open but not strictly OSI-approved OSS ™️  (e.g. a license like MIT, but with a clause saying we can revoke the app at any time)
- Keep MIT, but add CLAs to any OSS repo, it'd bug people the first time they submitted a PR and it'd give us the rights to change license / make private (which is what other companies do)

I guess the framing for the project creation should be useful here, like "start the debate from the perspective of open" rather than "start as open"

File: meta/engineering_principles.md
Comment (orta) [reply meta/engineering_principles.md]: Great point. 

We use the phrase "own" but it should be more metaphorical I think, for example we're not on the TypeScript core team - but we basically have their digital phone numbers. We probably will never need it, but we can easily make a meeting with them to discuss something critical.

I'll try shift the angle a bit here form a direct "repo access"

File: meta/engineering_principles.md
Comment (williardx) [inline meta/engineering_principles.md]: @joeyAghion I like the idea of "Done means done" as described but the name is a little ambiguous to me. When I read "done means done" I associate it with not returning to finished work while the description points more to following through.

File: meta/engineering_principles.md
Comment (williardx) [inline meta/engineering_principles.md]: I see two different but related concepts here: "own your dependencies" and "don't reinvent the wheel." You could in theory have one without the other but they go together pretty naturally. Should they be separate, or maybe described with a more general name?

Comment (orta) [reply meta/engineering_principles.md]: I think own your dependencies at core isn’t quite the same as don’t reinvent the wheel - this speaks to the acceptance that all the code in your stack is something you should contribute to

Don’t reinvent the wheel is a pretty good  way of thinking but it’s basically something most engineering teams should do - not sure if it is something unique enough to us to warrant inclusion

Comment (orta) [reply meta/engineering_principles.md]: Other than a sentence nod to it, probably

Comment (ashfurrow) [reply meta/engineering_principles.md]: Yeah, I see "don't reinvent the wheel" as the foundation that the "own your own dependencies" is built upon. A nod sounds like a good idea.

Comment (williardx) [reply meta/engineering_principles.md]: Agreed that "don't reinvent the wheel" should be baseline but in practice lots of teams end up reinventing the wheel because it's personally more gratifying, or because they believe their needs are unique enough to warrant it. That we take it seriously is kind of interesting!

Comment (ashfurrow) [reply meta/engineering_principles.md]: Yeah! In contrast to the "not invented here" idea, the objc.io article linked to describes the mobile team's ideology of "proudly discovered elsewhere", which I still admire.

File: meta/engineering_principles.md
Comment (joeyAghion) [reply meta/engineering_principles.md]: Interesting! I'd heard it this way so many times that the ambiguity disappeared for me, but I see what you're saying now.

I actually disagree with "not returning to finished work" so strongly that I'd like to suggest an additional principle: something like "Expect change." Thoughts?

File: meta/engineering_principles.md
Comment (orta) [reply meta/engineering_principles.md]: We should re-use that phrase here

File: meta/engineering_principles.md
Comment (orta) [reply meta/engineering_principles.md]: I feel like "coming back is a new project" could be a sub-part of explaining "done is done"

**Comment by orta** [comment]
I've taken a first run "build for 10x" and "done is done"

**Comment by orta** [comment]
@sweir27 mentioned "Being Nice is Nice" as one of our principals

**Comment by ashfurrow** [comment]
@orta @sweir27 Yeah yeah, I like that a lot. It's kind of... like, unspoken almost? Like, "being nice is nice" is just so much a part of how we work that it's easy to forget it's even there.

File: meta/engineering_principles.md
Comment (alloy) [inline meta/engineering_principles.md]: **Re**worked?

File: meta/engineering_principles.md
Comment (alloy) [inline meta/engineering_principles.md]: The first part of the second sentence should probably be part of the first sentence instead of the second.

File: meta/engineering_principles.md
Comment (alloy) [reply meta/engineering_principles.md]: It actually is as simple as ticking a box, because MIT licensed code may be used in closed source projects. The only thing that needs to happen is to not remove the license _or_ annotate the code from external contributors, which could easily be automated as that data exists in SCM anyways.

As for there being a new public parent repo, while that could happen it’s extremely unlikely to ever happen to a repo of ours and additionally isn’t really any different from people having local clones. It’s not something I would really consider in this debate.

File: meta/engineering_principles.md
Comment (sweir27) [inline meta/engineering_principles.md]: We could mention the fact that we intentionally try to allow teams/people to not silo from an organizational perspective by allowing ourselves to re-staff/move periodically.

**Comment by sweir27** [comment]
To flesh out the idea I had above (also have to give much credit to my mother who always says `"It's nice to be nice"` which is one of the more helpful life lessons).

---

"Being nice is nice" or "It's nice to be nice" or however we want to say it.

Behind every piece of code is a human, internal or external to Artsy. Give people the benefit of the doubt. Take the time to understand _why_ they made a decision before making assumptions. Realize that you may come across differently on slack/github than in-person and consider how the person on the other side might respond to what you're saying. There's _always_ a nice way to handle a situation, and we strive for that.

**Comment by orta** [comment]
Being nice is nice is good for me, I gave a talk on this [back in 2015](https://academy.realm.io/posts/altconf-orta-therox-being-nice-in-open-source/)

File: meta/engineering_principles.md
Comment (joeyAghion) [inline meta/engineering_principles.md]: My attempt at re-wording this (take what you like):

> Technology choices should strive to be "optimal" while avoiding over-engineering. When designing systems or evaluating scalability and performance, we aim for today's decisions to withstand 10x the traffic, data, or scale. This realistic horizon helps us balance the need to move quickly with the sometimes-competing need to invest in infrastructure and architecture. It also recognizes that solutions are expected to evolve and be replaced.

Comment (orta) [reply meta/engineering_principles.md]: I love it, replaced.

File: meta/engineering_principles.md
Comment (joeyAghion) [inline meta/engineering_principles.md]: My attempt:

> We take responsibility for not just implementing the defined work but ensuring it's the _right_ work, delivered all the way to [and understood by] the end consumer, instrumented, tested, and confirmed to behave as expected in production. This means we document, deploy, and monitor our changes. It's "done" when we can comfortably put it out of our minds and move on to new work.

**Comment by joeyAghion** [comment]
I like "being nice is nice" and think that some other "obvious" items might be worth making explicit for new team members:
* shipping early and often
* making the "hard" things easy with CI/CD/devops
* anybody can contribute to anything and our open-source-style workflow

**Comment by starsirius** [comment]
This is great! I like that we have examples and further readings for each section! Minor suggestion that it will help me read with a ToC type of summary on the top, given the # of items we have.

While reading through it, it reminds me of the triangle (quantity-quality-time). Maybe that's worth mentioning, too.

**Comment by mennenia** [comment]
❤️

**Comment by joeyAghion** [comment]
Another principle (from Platform) that felt relevant today as I was reviewing the last and next quarters: "Order matters." I.e., we think carefully about the right order in which to take on big projects. Initial steps can de-risk projects, expose mistaken assumptions, surface feedback, and create opportunities for reuse or other compounding benefits.

**Comment by orta** [comment]
OK, I think I've addressed all the feedback in here 

WRT the TOC, I'll fix and build something more systematic for the entire upcoming README repo rather than a one-off for this file.

**Comment by orta** [comment]
Will merge this on Tuesday if there's no breaking feedback, it's obviously a living document - I'd love some blog posts or links around the last two values so they feel fleshed out

**Comment by orta** [comment]
Alright, merging this. 🎉

**Pull request merged**
